scripts: check_init_priorities: use the zephyr executable file  [WAS: scripts: add dump_init_order.py]

fabiobaltieri opened the pull request:
Hi, this is a rework of  device discovery code of `check_init_priorities`. Instead of finding all the individual object files and find the level+piority from the section name, parse back the `initlevel` section from the output elf file. This means that the script is not impacted by stale files in the build directory, but more importantly it makes it work with LTO builds, where somehow the intermediate object fiels sections are completely mangled up.

While doing this, add a feature to print a human readable format of the initialization sections. This is pretty much just a friendlier version of `arm-zephyr-eabi-objdump -j initlevel -j device_area -d build/zephyr/zephyr.elf`. I'm using it to troubleshoot SYS_INIT order breakages, though I'm thinking one may use it in CI to compare the known sequence order of a build and catch unintended changes.

Looks something like this:

```
$ west build -t initlevels -p -b nrf52840dk_nrf52840 samples/drivers/led_pwm
...
EARLY
PRE_KERNEL_1
  __init_nordicsemi_nrf52_init: nordicsemi_nrf52_init(NULL)
  __init___device_dts_ord_74: clk_init(__device_dts_ord_74)
  __init_rtt_init: rtt_init(NULL)
  __init___device_dts_ord_11: gpio_nrfx_init(__device_dts_ord_11)
  __init___device_dts_ord_104: gpio_nrfx_init(__device_dts_ord_104)
  __init___device_dts_ord_112: uarte_0_init(__device_dts_ord_112)
  __init_uart_console_init: uart_console_init(NULL)
PRE_KERNEL_2
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init_enable_logger: enable_logger(NULL)
  __init_malloc_prepare: malloc_prepare(NULL)
  __init___device_dts_ord_69: pwm_nrfx_init(__device_dts_ord_69)
  __init___device_dts_ord_19: led_gpio_init(__device_dts_ord_19)
  __init___device_dts_ord_68: led_pwm_init(__device_dts_ord_68)
APPLICATION
SMP
```

Very much inspired by a comment from @bjarki-trackunit on the original proposal, should be fairly easy to adapt to https://github.com/zephyrproject-rtos/zephyr/pull/61986.

Also pushed on https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch

Since this now needs to work on a linked library I had to tweak the CMakeFiles a bit, and also it turns out this does not quite work with the elf file when using `NATIVE_LIBRARY`, although it does on the final exe. I could hack around it but I'd rather turn it off in that case right now and look for a proper solution down the road.

fabiobaltieri committed:
scripts: check_init_priorities: use the Zephyr executable file

Rework check_init_priorities to use the main executable file instead of
the individual object files for discovering the devices.

This should make the script more robust in case of stale files in the
build directory, and also makes it work with LTO object files.

Additionally, keep track of the detected init calls, and add a handy
"-i" option to produce a human readable print of the initcalls in the
call sequence, which can be useful for debugging initialization problems
due to odd SYS_INIT and DEVICE interactions.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

bjarki-andreasen reviewed (dismissed):
Love it :)

kartben commented:
Very cool! Should this be added as a west target for convenience/ better discoverability?

fabiobaltieri commented:
Thanks for the feedback folks, I thought more about this and  realized that the information here works for `check_init_priorities.py`, and does not require the hack for detecting sub builds plus it works with LTO files (which turns out the corrent code does not).

So I'm taking a 180 on this one, going to rip the file scanning code off `check_init_priorities.py`, replace it with this one and I'll integrate the functionality of this `check_init_priorities.py` itself.

@kartben yes will look into adding a target

fabiobaltieri committed:
cmake: add a new "initlevels" target to print the init sequence

Add a new "initlevels" target that can be used to print a human readable
version of the init sequence by running: west build -t initlevels.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Update: this is now a `check_init_priorities` update with the "dump initlevels" as a feature and cmake target. Ran it on CI as well, ended excluding it when using `CONFIG_NATIVE_LIBRARY` (i.e. natsim) to avoid making a giant mess with CMake to make it run on the right binary at the right time but I'll look into fixing that down the road.

gmarull reviewed (commented):
just one comment: since this is now using the Zephyr elf file, shouldn't we re-use/extend `elf_parser.py`?

fabiobaltieri commented:
> just one comment: since this is now using the Zephyr elf file, shouldn't we re-use/extend `elf_parser.py`?

Yeah I initially started building on top of it and eventually realized that there was going to be pretty much no overlap. Despite the generic sounding name that's very much specialzied on processing the dependency graph, I found I was making a mess of it with no deduplication savings so I went with a separate container class.

nordicjm commented on scripts/build/check_init_priorities.py:
```diff
@@ -149,35 +168,64 @@ def _device_ord_from_rel(self, rel):
+        elif elfclass == 64:
+            ptrsize = 8
+        else:
+            ValueError(f"Uknown pointer size for ELF class f{elfclass}")
```
*Unknown

nordicjm reviewed (commented):
Minor nit. but looks good!

henrikbrixandersen reviewed (commented):
Minor nit: "zephyr" is misspelled in the commit message.

fabiobaltieri commented:
> Minor nit: "zephyr" is misspelled in the commit message.

Yeah found a couple more typos as well in there :-) sorry... fixed

henrikbrixandersen reviewed (dismissed):
LGTM, thanks!

fabiobaltieri commented:
Thanks folks, sorry about the typos, need to remember doing a spell checker pass on my commit and error messages.

tejlmand reviewed (approved):
nothing wrong, but gotta test this out with armclang before final approval.

tejlmand reviewed (changes requested):
nothing wrong, but gotta test this out with armclang before final approval.

tejlmand commented on CMakeLists.txt:
```diff
@@ -1811,12 +1811,20 @@ if(CONFIG_CHECK_INIT_PRIORITIES)
     COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/build/check_init_priorities.py
-      --build-dir ${PROJECT_BINARY_DIR}/..
-      --edt-pickle ${EDT_PICKLE}
+      --elf-file=${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME}
```
Failing with `armclang` with following error:
```
Traceback (most recent call last):
  File "/projects/github/ncs/zephyr/scripts/build/check_init_priorities.py", line 394, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/projects/github/ncs/zephyr/scripts/build/check_init_priorities.py", line 376, in main
    validator = Validator(args.elf_file, args.edt_pickle, log)
  File "/projects/github/ncs/zephyr/scripts/build/check_init_priorities.py", line 253, in __init__
    self._obj = ZephyrInitLevels(elf_file_path)
  File "/projects/github/ncs/zephyr/scripts/build/check_init_priorities.py", line 110, in __init__
    self._load_level_addr()
  File "/projects/github/ncs/zephyr/scripts/build/check_init_priorities.py", line 145, in _load_level_addr
    raise ValueError(f"Missing init symbols, found: {self._init_level_addr}")
ValueError: Missing init symbols, found: {}
ninja: build stopped: subcommand failed.
```

fabiobaltieri replied:
Bummer. Would you be able to send an elf file over to me?

gmarull replied:
I'd honestly not block this PR because it doesn't work with armclang, but exclude the target if the toolchain being used is armclang. It can be investigated later by users with access to it. Restricting progress on OSS projects because of licensed software is not a good idea.

nashif replied:
does it work with public clang?

fabiobaltieri replied:
```
$ export ZEPHYR_TOOLCHAIN_VARIANT=llvm
$ west build -p -b native_posix_64 samples/subsys/display/lvgl
...
-- Found toolchain: host (clang/ld)
...
-- The C compiler identification is Clang 14.0.6
-- The CXX compiler identification is Clang 14.0.6
-- The ASM compiler identification is Clang with GNU-like command-line
...
[329/329] Linking C executable zephyr/zephyr.elf
$ ./scripts/build/check_init_priorities.py -v
INFO: check_init_priorities: build/zephyr/zephyr.elf
INFO: /gpio@800/sdl_gpio POST_KERNEL 3 > /gpio@800 POST_KERNEL 2
INFO: /keys POST_KERNEL 5 > /gpio@800 POST_KERNEL 2
INFO: /keys POST_KERNEL 5 > /gpio@800 POST_KERNEL 2
INFO: /lvgl_button_input APPLICATION 2 > /keys POST_KERNEL 5
INFO: /lvgl_pointer APPLICATION 3 > /input-sdl-touch POST_KERNEL 6
$ ./scripts/build/check_init_priorities.py -i
EARLY
PRE_KERNEL_1
  __init_statics_init_pre: statics_init(NULL)
  __init___device_dts_ord_18: np_uart_0_init(__device_dts_ord_18)
  __init_uart_console_init: uart_console_init(NULL)
  __init_posix_arch_console_init: posix_arch_console_init(NULL)
PRE_KERNEL_2
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init_enable_logger: enable_logger(NULL)
  __init_enable_shell_uart: enable_shell_uart(NULL)
  __init___device_dts_ord_31: gpio_emul_init(__device_dts_ord_31)
  __init___device_dts_ord_32: gpio_sdl_init(__device_dts_ord_32)
  __init_k_sys_work_q_init: k_sys_work_q_init(NULL)
  __init___device_dts_ord_10: gpio_keys_init(__device_dts_ord_10)
  __init___device_dts_ord_12: sdl_init(__device_dts_ord_12)
APPLICATION
  __init___device_dts_ord_16: sdl_display_init(__device_dts_ord_16)
  __init_lvgl_init: lvgl_init(NULL)
  __init___device_dts_ord_11: lvgl_button_input_init(__device_dts_ord_11)
  __init___device_dts_ord_13: lvgl_pointer_input_init(__device_dts_ord_13)
SMP
```

Yeah looks alright to me, the armglang linker must be discarding the symbols, maybe the the binary is stripped? Hard to say without a file...

tejlmand reviewed (changes requested):
doesn't work with elf file produced by armclang.

need more investigation.

fabiobaltieri commented:
Ok small update now that I managed to get my hands on an armclang file, there's two problems:

1 - the init data is spread into multiple sections, looks something like:
```
        0x000021d8 - 0x00002200 is init_103
        0x00002200 - 0x00002208 is init_104
        0x00002208 - 0x00002208 is init_end
```
2 - the symbols for the init section boundaries are not there, looks like the linker resolved and ate them up

So [1] is easy to fix and I've actually decided to integrate the change regardless: I've just changed `_initlevel_pointer` to resolve the section for every pointer individually, which is actually nice regardless cause it means the script does not need to look for a specific section name to find the data.

[2] is a more of a problem though, if the armclang linker cannot be configured to omit those symbols (some debug option?) I think I'll have to either guess the symbols from those `init_103` strings (is that dependable?!) or just disable this for armclang entirely.

Waiting for further feedback from @tejlmand.

fabiobaltieri commented:
> ```
>         0x000021d8 - 0x00002200 is init_103
>         0x00002200 - 0x00002208 is init_104
>         0x00002208 - 0x00002208 is init_end
> ```

Wonder how the linker ended up with those now that I looked into it, from the content these seems to be 103: `PRE_KERNEL_1` and 104: `PRE_KERNEL_2`.

Was trying to correlate those in the code but maybe it's better to just shut this off for armclang for the moment. Waiting to hear if it works for ARCMWDT at least.

fabiobaltieri committed:
cmake: skip build time priority checking with armclang

The script does not play well with armclang binariest at the moment.
Disable the automatic invocation when running with armclang, at least
until this is investigated and fixed.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Added a conditional to skip this for armclang binaries, no point blocking on that, it can be fixed down the road by someone that have access to that compiler.

evgeniy-paltsev reviewed (dismissed):
Tested with MWDT toolchain - seems to be working fine.

fabiobaltieri commented:
@evgeniy-paltsev thanks for testing this

@tejlmand can we go ahead with this one excluded on armclang for now? Then you folks with access to that compiler can look into this whenever you have time

tejlmand commented on CMakeLists.txt:
```diff
@@ -1816,6 +1816,15 @@ if(CONFIG_CHECK_INIT_PRIORITIES)
 endif()
 
+add_custom_target(
+  initlevels
```
we should not introduce a build target which will fail in some setups when executed by the user.

In such cases, creation of the build target should be properly safe guarded.

fabiobaltieri replied:
Got it, guarded, kept it in a separate commit, easy to rollback if that gets fixed.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -747,6 +747,7 @@ config BUILD_OUTPUT_STRIP_PATHS
 config CHECK_INIT_PRIORITIES
 	bool "Build time initialization priorities check"
 	default y
+	depends on !NATIVE_LIBRARY
```
nothing wrong on this line, but could you add a safe-guard here, so that this not available for armclang.

Something like:
```suggestion
	depends on !NATIVE_LIBRARY
	depends on "$(ZEPHYR_TOOLCHAIN_VARIANT)" != "armclang"
```

fabiobaltieri replied:
Aha, thanks, sorry somehow I could not find a Kconfig option for this, that works.

tejlmand commented on CMakeLists.txt:
```diff
@@ -1804,7 +1804,8 @@ if(CONFIG_BUILD_OUTPUT_INFO_HEADER)
 
-if(CONFIG_CHECK_INIT_PRIORITIES)
+if(CONFIG_CHECK_INIT_PRIORITIES AND
+   NOT CMAKE_C_COMPILER_ID STREQUAL "ARMClang")
```
we try to make the general CMake build system as independent on compilers / os / etc as possible.

Toolchain specifics should try to be kept in dedicated cmake files for such purposes, such as those under `cmake/toolchain` / `cmake/compiler` so rest of CMake code is as common as possible.

Therefore better to ensure the Kconfig option is not available to users using such toolchain (see other comment).

fabiobaltieri replied:
Sounds good, would have been my option as well but somehow I missed the Kconfig option when I looked for it, dropped the check here, but left it on the target as that's independent from the option.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -757,6 +757,9 @@ config CHECK_INIT_PRIORITIES
 	  priority), see CHECK_INIT_PRIORITIES_FAIL_ON_WARNING to fail on
 	  warnings (dependent devices, same priority) as well.
 
+	  NOTE: this check is does not support armclang binaries and is
+	  currently silently skipped for those.
+
```
if adding dependency on setting, then this note can be removed.
```suggestion
```

fabiobaltieri replied:
Yes! done.

tejlmand reviewed (changes requested):
looks good, just a couple of minor nits to get cleaned out before final +1.

fabiobaltieri commented:
Reworked the "no armclang support" patch, it now works by:
- not defining the `initlevels` target
- adding a Kconfig dep instead

tejlmand commented on CMakeLists.txt:
```diff
@@ -1811,12 +1811,22 @@ if(CONFIG_CHECK_INIT_PRIORITIES)
     )
 endif()
 
+if(NOT CMAKE_C_COMPILER_ID STREQUAL "ARMClang")
```
not fond, of it, but I see your reasoning here.
https://github.com/zephyrproject-rtos/zephyr/pull/62459#discussion_r1331431765

and users could want to run the target even with the kconfig to `n`.
I don't think this specific feature is strong enough to have something along `TOOLCHAIN_CREATES_GOOD_ELF` or similar.

So accepted for now, and let's see if we can find a way with armclang to support the check init feature.

fabiobaltieri replied:
Yeah, I'm sure it can be fixed, those stripping options in `cmake/bintools/armclang/elfconvert_command.cmake` looks suspicious to me. It'd be cool to figure a way of having a CI run with these compilers so we don't depend on individuals to test stuff though.
